types/react/jsx-dev-runtime.d.ts

```diff
@@ -0,0 +1,3 @@
+import { createElement } from './';
```

Comment:
Those files should export `JSX` namespace - you can look at how I have done it for Emotion here: https://github.com/emotion-js/emotion/pull/2078/files . 

It would probably help if the `JSX` namespace in the `index` would actually be exported somehow rather than just attached to the global namespace. I've attempted to do a similar refactor here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46351 but I didn't receive the required approvals. You can maybe incorporate part of this work into your branch.

Reply:
Can we fix this in a follow-up? The current approach at least fixed https://github.com/facebook/create-react-app/issues/10109 and has less risk of breaking existing patterns. As far as I know the jsx `jsx-runtime` and `createElement` allow are the same structurally.

Reply:
> Those files should export JSX namespace - you can look at how I have done it for Emotion here: emotion-js/emotion#2078 (files) .

As far as I can tell `import './index'` works as well since that `./index` adds the JSX namespace globally.

This seems safer to me and any refactoring we want to do to how we handle JSX namespaces can be done in a separate PR. Or would this block emotion? If so could you propose a change to this branch so that I can understand what exactly you're trying to do?

Reply:
> Can we fix this in a follow-up? The current approach at least fixed facebook/create-react-app#10109 and has less risk of breaking existing patterns.

My intuition (but I haven't fully confirmed it) is that it's exactly what you have mentioned in the later post - `import './index'` does the trick and not the `createElement` part. So I'm just proposing to sticking to the first one because exporting `createElement` is not correct here

> As far as I know the jsx jsx-runtime and createElement allow are the same structurally.

Not really - `createElement` accepts variadic length of parameters while `jsx` and `jsxs` have fixed length of those. Those signatures do not match.

> Or would this block emotion?

Can't be totally sure, because how TS resolves those things is magic to me. My bet is that it doesn't affect us much because the global `JSX` namespace should be overshadowed by the one that we provide - so I think this is OK.

OTOH I believe that there is no need at all to provide a global JSX namespace and it's probably an artifact of the past and this (new major version of those types here) could be a good moment to clean this up.

Reply:
> Not really - createElement accepts variadic length of parameters while jsx and jsxs have fixed length of those. Those signatures do not match.

I'm specifically talking about the JSX. Not how it's transpiled. Since you don't "manually call"  `jsx` I don't see an issue.

> OTOH I believe that there is no need at all to provide a global JSX namespace and it's probably an artifact of the past and this (new major version of those types here) could be a good moment to clean this up.

I'm not doing feature changes and refactorings of existing systems in the same PR. Especially if I'm not aware of a concrete issue the current approach has. 

So it would help a lot if we talk about concrete issues (ideally that we can reproduce) instead of theoretical concerns. There are plenty of issues with the react typings that we didn't fix with 17 because we were [explicitly asked to keep 17 free from breaking changes](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/46691#issuecomment-672840456). And that's my goal: provide typings for React 17 that don't interfere with React.

Reply:
> I'm specifically talking about the JSX. Not how it's transpiled. Since you don't "manually call" jsx I don't see an issue.

That's fine by me and yes - JSX namespace works the same for both runtimes. My sole point is that this `createElement` should not appear in those new entries because it doesn't match reality and I don't think it's solving anything (unless it is but I somewhat doubt it - I've been tinkering lately with fixing JSX namespace lookup in TS and I don't see how this would be required here, but again - the whole thing is quirky to me so I might be off about some details).

> So it would help a lot if we talk about concrete issues (ideally that we can reproduce) instead of theoretical concerns. There are plenty of issues with the react typings that we didn't fix with 17 because we were explicitly asked to keep 17 free from breaking changes. And that's my goal: provide typings for React 17 that don't interfere with React.

👍